hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/impl/connector/kafka/StreamKafkaP.java

```diff
@@ -91,53 +96,64 @@
     protected void init(@Nonnull Context context) throws Exception {
         snapshottingEnabled = context.snapshottingEnabled();
         consumer = new KafkaConsumer<>(properties);
-        reassignPartitions();
+        reassignPartitions(true);
     }
 
-    private void reassignPartitions() {
-        boolean changed = false;
+    private void reassignPartitions(boolean isInitial) {
+        List<TopicPartition> addedPartitions = new ArrayList<>();
         // check for added partitions (kafka doesn't support partition removal). Initially, all partitions are added.
         for (int topicIdx = 0; topicIdx < topicIds.size(); topicIdx++) {
             String topicName = topicIds.get(topicIdx);
             List<PartitionInfo> partitionInfos = consumer.partitionsFor(topicName);
-            getLogger().finest("Num of ptions for topic " + topicName + ": " + partitionInfos.size());
+            LoggingUtil.logFinest(getLogger(), "Num of ptions for topic '%s': %d", topicName, partitionInfos.size());
+            if (nextAssignablePtions[topicIdx] >= partitionInfos.size()) {
+                continue;
+            }
+            // Sort partitionInfos
+            // Kafka returns internal array in partitionsFor(), so I'd better do a copy first to avoid problems
+            partitionInfos = new ArrayList<>(partitionInfos);
+            partitionInfos.sort(comparing(PartitionInfo::partition));
             while (nextAssignablePtions[topicIdx] < partitionInfos.size()) {
                 int partition = nextAssignablePtions[topicIdx];
                 nextAssignablePtions[topicIdx] += processorCount;
                 assert partitionInfos.get(partition).partition() == partition;
-                assignment.add(new TopicPartition(topicName, partition));
-                changed = true;
+                addedPartitions.add(new TopicPartition(topicName, partition));
             }
         }
 
-        if (changed) {
-            getLogger().info("Partition assignment changed: " + assignment);
+        if (!addedPartitions.isEmpty()) {
+            getLogger().info("Partition assignment changed, new partitions: " + addedPartitions);
+            assignment.addAll(addedPartitions);
             consumer.assign(assignment);
+            if (!isInitial) {
+                consumer.seekToBeginning(addedPartitions);
```

Comment:
does this call take `auto.offset.reset` into account?

Reply:
Intentionally not. This is for partitions added after the initial assignment: we want to read them from the beginning.